### PR TITLE
[runtimes] Always define cxx_shared, cxx_static & other targets

### LINKED ISSUE #77654: [D134221] [runtimes] Always define the cxx_shared, cxx_static & other targets
This issue tracks picking up https://reviews.llvm.org/D134221 from the Phabricator archive.

### PR DESCRIPTION
**Pull request by ldionne**
This patch always defines the cxx_shared, cxx_static & other top-level targets. However, they are marked as EXCLUDE_FROM_ALL when we don't want to build them. Simply declaring the targets should be of no harm, and it allows other projects to mention these targets regardless of whether they end up being built or not.

This patch basically moves the definition of e.g. cxx_shared out of the `if (LIBCXX_ENABLE_SHARED)` and instead marks it as EXCLUDE_FROM_ALL conditionally on whether LIBCXX_ENABLE_SHARED is passed. It then does the same for libunwind and libc++abi targets. I purposefully avoided to reformat the files (which now has inconsistent indentation) because I wanted to keep the diff minimal, and I know this is an area of the code where folks may have downstream diffs. I will re-indent the code separately once this patch lands.

This is a reapplication of 79ee0342dbf0, which was reverted in a3539090884c because it broke the TSAN and the Fuchsia builds.

Resolves #77654

Differential Revision: https://reviews.llvm.org/D134221

### CODE REVIEW

**Comment by ldionne** [comment]
CC @petrhosek @vitalybuka since this broke your setups last time. The reproduction steps were really non-trivial so I never got around to reproducing, but it would be great to see if this is still problematic.

**Comment by ldionne** [comment]
@petrhosek @vitalybuka Gentle ping.

If the CI is green, I will land this next week unless someone objects to it. I don't want to purposefully break anyone's build, but I also don't want to hold off on this change forever if there's nothing actionable.

Last time in https://reviews.llvm.org/D134221#3853050, @petrhosek you said:

> All Fuchsia builders started failing after this change landed with the following error:
> 
> ```
> /b/s/w/ir/x/w/staging/llvm_build/./bin/clang++ --target=x86_64-unknown-fuchsia --sysroot=/b/s/w/ir/x/w/cipd/sdk/arch/x64/sysroot -fPIC --target=x86_64-unknown-fuchsia -I/b/s/w/ir/x/w/cipd/sdk/pkg/sync/include -I/b/s/w/ir/x/w/cipd/sdk/pkg/fdio/include -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -ffunction-sections -fdata-sections  -O3 -DNDEBUG  -L/b/s/w/ir/x/w/cipd/sdk/arch/x64/lib -Wl,-z,defs  -nostdlib++ -shared -Wl,-soname,libc++abi.so.1 -o lib/libc++abi.so.1.0 libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_aux_runtime.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_default_handlers.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_demangle.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_exception_storage.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_guard.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_handlers.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_vector.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_virtual.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/stdlib_exception.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/stdlib_stdexcept.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/stdlib_typeinfo.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/abort_message.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/fallback_malloc.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/private_typeinfo.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/stdlib_new_delete.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_noexception.cpp.obj libcxxabi/src/CMakeFiles/cxxabi_shared_objects.dir/cxa_thread_atexit.cpp.obj  -lc  -lgcc  -lgcc_s && :
> ld.lld: error: unable to find library -lgcc
> ld.lld: error: unable to find library -lgcc_s
> ```
> There's no libgcc on Fuchsia, we always use compiler-rt builtins so it's not clear to me why the build is now trying link libgcc when targeting Fuchsia after this change.
> 

It looks like this *could* be caused by a check for something like `if (TARGET cxxabi_shared)` that used to be false and is now true, but it's hard to say without more information (and ideally reproducing locally).

**Comment by vitalybuka** [comment]
Can't comment on fuchsia, but buildbot_standard.sh which is mentioned on original review was shutdown for a while.
Some steps were moved to buildbot_cmake.sh based bots, I can try those.

**Review by vitalybuka — APPROVED** [review]
I can not reproduce linux bot issue with the patch.

**Comment by ldionne** [comment]
I had to fix how we were excluding targets from `EXCLUDE_FROM_ALL`, which didn't work as intended and was breaking the MinGW CI. I'll merge this if the CI passes.

**Comment by petrhosek** [comment]
I tested this locally in a clean build but I'm seeing the following build error:
```
CMake Error:
  Running

   '/usr/local/google/home/phosek/fuchsia/prebuilt/third_party/ninja/linux-x64/ninja' '-C' '/usr/local/google/home/phosek/llvm/llvm-project/build/fuchsia/runtimes/runtimes-armv6m-none-eabi-bins' '-t' 'recompact'

  failed with:

   ninja: error: build.ninja:52568: multiple rules generate /usr/local/google/home/phosek/llvm/llvm-project/build/fuchsia/lib/armv6m-unknown-none-eabi/libc++.a
```

**Comment by ldionne** [comment]
@petrhosek Where is the cache generating that build?

The fix for this is:

```
# On embedded platforms that don't support shared library targets, CMake implicitly changes shared
# library targets to be static library targets. This results in duplicate definitions of the static
# library targets even though we might not ever build the shared library target, which breaks the
# build. To work around this, we change the output name of the  shared library target so that it
# can't conflict with the static library target.
#
# This is tracked by https://gitlab.kitware.com/cmake/cmake/-/issues/25759.
set(LIBCXX_SHARED_OUTPUT_NAME "c++-shared" CACHE STRING "")
set(LIBCXXABI_SHARED_OUTPUT_NAME "c++abi-shared" CACHE STRING "")
set(LIBUNWIND_SHARED_OUTPUT_NAME "unwind-shared" CACHE STRING "")
```

**Comment by petrhosek** [comment]
> @petrhosek Where is the cache generating that build?
> 
> The fix for this is:
> 
> ```
> # On embedded platforms that don't support shared library targets, CMake implicitly changes shared
> # library targets to be static library targets. This results in duplicate definitions of the static
> # library targets even though we might not ever build the shared library target, which breaks the
> # build. To work around this, we change the output name of the  shared library target so that it
> # can't conflict with the static library target.
> #
> # This is tracked by https://gitlab.kitware.com/cmake/cmake/-/issues/25759.
> set(LIBCXX_SHARED_OUTPUT_NAME "c++-shared" CACHE STRING "")
> set(LIBCXXABI_SHARED_OUTPUT_NAME "c++abi-shared" CACHE STRING "")
> set(LIBUNWIND_SHARED_OUTPUT_NAME "unwind-shared" CACHE STRING "")
> ```

With #111791 everything builds without any issues.

**Comment by ilovepi** [comment]
Hi, I think we're seeing a bunch of ASAN test failures both on Aarch64 and x86_64 Linux bots. It seems like these tests are now failing to find the right shared library. Would you mind taking a look?

 @petrhosek do you know if this is something we need to update in our cache file? or is this a more general problem w/ the CMake?

Aarch64 bot: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-arm64/b8734460807473127393/overview
x64 bot: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8734465123369888721/overview

Logs for x64: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8734465123369888721/+/u/clang/test/stdout

Example error:
```
FAIL: AddressSanitizer-x86_64-linux :: TestCases/Linux/asan_default_suppressions.cpp (50 of 15170)
******************** TEST 'AddressSanitizer-x86_64-linux :: TestCases/Linux/asan_default_suppressions.cpp' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 2: /b/s/w/ir/x/w/llvm_build/./bin/clang  --driver-mode=g++ -fsanitize=address -mno-omit-leaf-frame-pointer -fno-omit-frame-pointer -fno-optimize-sibling-calls -gline-tables-only   -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta   /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/asan/TestCases/Linux/asan_default_suppressions.cpp -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/asan/X86_64LinuxConfig/TestCases/Linux/Output/asan_default_suppressions.cpp.tmp && not  /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/asan/X86_64LinuxConfig/TestCases/Linux/Output/asan_default_suppressions.cpp.tmp 2>&1 | FileCheck /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/asan/TestCases/Linux/asan_default_suppressions.cpp
+ /b/s/w/ir/x/w/llvm_build/./bin/clang --driver-mode=g++ -fsanitize=address -mno-omit-leaf-frame-pointer -fno-omit-frame-pointer -fno-optimize-sibling-calls -gline-tables-only -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/asan/TestCases/Linux/asan_default_suppressions.cpp -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/asan/X86_64LinuxConfig/TestCases/Linux/Output/asan_default_suppressions.cpp.tmp
+ not /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/asan/X86_64LinuxConfig/TestCases/Linux/Output/asan_default_suppressions.cpp.tmp
+ FileCheck /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/asan/TestCases/Linux/asan_default_suppressions.cpp
/b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/asan/TestCases/Linux/asan_default_suppressions.cpp:6:11: error: CHECK: expected string not found in input
// CHECK: AddressSanitizer: failed to parse suppressions
          ^
<stdin>:1:1: note: scanning from here
/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/asan/X86_64LinuxConfig/TestCases/Linux/Output/asan_default_suppressions.cpp.tmp: error while loading shared libraries: libc++.so.2: cannot open shared object file: No such file or directory
^
<stdin>:1:116: note: possible intended match here
/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/asan/X86_64LinuxConfig/TestCases/Linux/Output/asan_default_suppressions.cpp.tmp: error while loading shared libraries: libc++.so.2: cannot open shared object file: No such file or directory
                                                                                                                   ^

Input file: <stdin>
Check file: /b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/test/asan/TestCases/Linux/asan_default_suppressions.cpp

-dump-input=help explains the following input dump.

Input was:
<<<<<<
           1: /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/test/asan/X86_64LinuxConfig/TestCases/Linux/Output/asan_default_suppressions.cpp.tmp: error while loading shared libraries: libc++.so.2: cannot open shared object file: No such file or directory 
check:6'0     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:6'1                                                                                                                        ?                                                                                                                                                                     possible intended match
           2: error: No such file or directory 
check:6'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>

--

********************
```

**Comment by ldionne** [comment]
@ilovepi Some things that *could* be going wrong: The `cxx_shared` target is now defined when it wasn't before, and something in your CMake has e.g. `if (TARGET cxx_shared)` and does something it shouldn't based on that.

Otherwise, it's also possible that `libc++.so` doesn't get installed at the right location anymore in case it's not part of the `all` target under that configuration, but that seems unlikely because previously the target would not have existed at all.

**Comment by ldionne** [comment]
Also, this is another reminder that we would benefit from having Fuchsia builders in the pre-commit CI. We've asked before and we'll keep asking until we get them: Fuchsia is amongst the last platform that frequently reports errors but doesn't provide us with any pre-commit visibility.

**Comment by ilovepi** [comment]
I think I understand what's happening, but not why. We don't normally build `cxx_shared` on Linux, but now we do. I'm not sure why that's causing a problem on this test per se, but we set `LIBCXX_ENABLE_SHARED=OFF` in our build files, so I'm surprised its happening now. In fact under `build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins` I only see `LIBCXX_ENABLE_SHARED=OFF` in all of the CMake files. Seems like this isn't propagating correctly anymore?

**Comment by ilovepi** [comment]
> Also, this is another reminder that we would benefit from having Fuchsia builders in the pre-commit CI. We've asked before and we'll keep asking until we get them: Fuchsia is amongst the last platform that frequently reports errors but doesn't provide us with any pre-commit visibility.

Agreed. I'd also like us to be in that situation.

File: libcxx/src/CMakeLists.txt
Comment (ilovepi) [inline libcxx/src/CMakeLists.txt:277]: Is this reversed? https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:IF seems to indicate that the order should be `True, False`. Or does the logic need to be inverted for `EXCLUDE_FROM_ALL`?

Comment (ilovepi) [reply libcxx/src/CMakeLists.txt:277]: nvmd I see my mistake. Disregard.

**Comment by MarkMurrayARM** [comment]
Hi - is this being worked on? We are still seeing build failures that look like
```
-- ABI list file not generated for configuration aarch64-none-elf.libcxxabi.v2.unstable.exceptions.nonew, `check-cxx-abilist` will not be available.
-- Configuring done
-- Generating done
CMake Error:
  Running

   '/usr/bin/ninja' '-C' '/home/marmur02/oss/LLVM-embedded-toolchain-for-Arm/build/libcxx_libcxxabi_libunwind/aarch64a_exn_rtti/src/libcxx_libcxxabi_libunwind_aarch64a_exn_rtti-build' '-t' 'recompact'

  failed with:
   ninja: error: build.ninja:524: multiple rules generate lib/libunwind.a [-w dupbuild=err]
```
... which don't happen with this change locally backed out.

**Comment by MarkMurrayARM** [comment]
NM - fixed in our builds.

**Comment by ldionne** [comment]
If other folks get here because of an error like `multiple rules generate lib/SOMETHING.a`, this is the issue (from https://github.com/llvm/llvm-project/pull/80007#issuecomment-2402188675):

> 
> The fix for this is:
> 
> ```
> # On embedded platforms that don't support shared library targets, CMake implicitly changes shared
> # library targets to be static library targets. This results in duplicate definitions of the static
> # library targets even though we might not ever build the shared library target, which breaks the
> # build. To work around this, we change the output name of the  shared library target so that it
> # can't conflict with the static library target.
> #
> # This is tracked by https://gitlab.kitware.com/cmake/cmake/-/issues/25759.
> set(LIBCXX_SHARED_OUTPUT_NAME "c++-shared" CACHE STRING "")
> set(LIBCXXABI_SHARED_OUTPUT_NAME "c++abi-shared" CACHE STRING "")
> set(LIBUNWIND_SHARED_OUTPUT_NAME "unwind-shared" CACHE STRING "")
> ```

Also, please consider pinging the CMake issue referenced.

**Pull request merged**
